Add initial files.

tolson4 opened the pull request.

tolson4 committed:
Add initial files.

tolson4 committed:
Cleanup and reorganize testing app.

tolson4 committed:
Cleanup, convert image component to vue3.

tolson4 committed:
Add github actions workflows.

tolson4 committed:
Fix linting errors.

tolson4 committed:
Cleanup.

tolson4 committed:
Remove prefix.

davidlehn committed:
Use test karma coverage commands.

davidlehn committed:
Update eslint support.

- Update eslint packages.
- Lint module.

davidlehn committed:
Add changelog.

davidlehn committed:
Update dependencies.

davidlehn committed:
Update files field.

dlongley commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,79 @@
+    timeout-minutes: 10
+    services:
+      mongodb:
+        image: mongo:4.2
```
It doesn't look like mongodb is used here, so we should remove needing the service from CI.

tolson4 replied:
7d1067c36fc326d38ac69cd676f4599ab88736af

dlongley commented on LICENSE.md:
```diff
@@ -0,0 +1,115 @@
+Bedrock Non-Commercial License v1.0
+===================================
+
+Copyright (c) 2011-2021 Digital Bazaar, Inc.
```
```suggestion
Copyright (c) 2011-2022 Digital Bazaar, Inc.
```

dlongley commented on components/CredentialBase.vue:
```diff
@@ -0,0 +1,157 @@
+/*!
+ * Copyright (c) 2018-2022 Digital Bazaar, Inc. All rights reserved.
+ */
+import DynamicImage from './DynamicImage.vue';
+import CredentialField from './CredentialField.vue';
+import CredentialSwitch from './CredentialSwitch.vue';
+import {computed, defineProps, reactive, toRef, unref} from 'vue';
+import {useCredentialCommon} from './credentialCommon.js';
```
These should be sorted ... @davidlehn can you help with the eslint rule here?

tolson4 replied:
Sorted here: 4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialBase.vue:
```diff
@@ -0,0 +1,157 @@
+  console.log(state.details);
+  state.details = !state.details;
+}
+
```
```suggestion
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialDetail.vue:
```diff
@@ -0,0 +1,52 @@
+</script>
+
+<style lang="scss" scoped>
+
```
Here and anywhere else:

```suggestion
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialDetail.vue:
```diff
@@ -0,0 +1,52 @@
+} = useCredentialCommon({
+  credential: toRef(props, 'credential')
+});
+
```
Here and anywhere else:

```suggestion
```

tolson4 replied:
eb427522ee1031f96a5f346e92d4afe6a897149e

dlongley commented on components/CredentialField.vue:
```diff
@@ -0,0 +1,54 @@
+</template>
+
+<script setup>
+import {computed, defineProps} from 'vue';
```
Missing copyright header here and anywhere else.

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialField.vue:
```diff
@@ -0,0 +1,54 @@
+
+const show = computed(() => {
+  return props.value.length > 0;
+});
```
Could be simplified:

```suggestion
const show = computed(() => props.value.length > 0);
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+  const map = {};
+  for(const registration of registrations) {
+    const {acceptableTypes, components} = registration;
+    //loop through types so that map will consist of an object
+    //containing all registered types
```
```suggestion
    // loop through types so that map will consist of an object
    // containing all registered types
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+    //loop through types so that map will consist of an object
+    //containing all registered types
+    for(const type of acceptableTypes) {
+      //If the type already exists in the map, merge it with the
+      //new information
```
```suggestion
      // if the type already exists in the map, merge it with the
      // new information
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+
+const selection = computed(() => {
+  const {credentialDisplay: {registration: registrations}} = config;
+  const map = {};
```
Use a `Map` instance instead. It's more performant and doesn't result in creating N-many hidden compiled classes in v8 (N = number of properties added).

tolson4 replied:
eb427522ee1031f96a5f346e92d4afe6a897149e

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+  const {mode, credential} = props;
+  const {type = ''} = unref(credential);
+  //work backwards to lookup type in mapping
+  //use defaults if not found in mapping
```
Here and anywhere else -- always put a space after `//`.

```suggestion
  // work backwards to lookup type in mapping
  // use defaults if not found in mapping
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+      return componentDefaults[mode];
+    }
+    const granularType = type[index];
+    if(map[granularType]?.[mode]) {
+      options = map[granularType][mode];
+    }
```
Only hit the map once:

```suggestion
    options = map.get(granularType)?.[mode];
```

tolson4 replied:
eb427522ee1031f96a5f346e92d4afe6a897149e

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+  // selection to walk through the options and select the first
+  // available component for the type.
+  const {component} = options[0];
+  return `${component}`;
```
Is component always already a string? If so:

```suggestion
  return component;
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -0,0 +1,85 @@
+import {config} from '@bedrock/web';
+import CredentialBase from './CredentialBase.vue';
+import CredentialDetail from './CredentialDetail.vue';
+import {merge} from 'lodash-es';
```
We've tried really hard to remove `lodash` from everything because of its various security maintenance issues. Is there a simpler local solution that can be used here -- or something else we could use like `extend` from `@bedrock/web`?

tolson4 replied:
eb427522ee1031f96a5f346e92d4afe6a897149e

dlongley commented on components/DynamicImage.vue:
```diff
@@ -0,0 +1,164 @@
+
+const showDefault = computed(() => {
+  return !props.src;
+});
```
```suggestion
const showDefault = computed(() => !props.src);
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/DynamicImage.vue:
```diff
@@ -0,0 +1,164 @@
+  return !props.src;
+});
+
+const defaultImage = computed(() => {
+  return $q.iconSet.credentialCard.defaultImage;
+});
```
```suggestion
const defaultImage = computed(
  () => $q.iconSet.credentialCard.defaultImage);
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/DynamicImage.vue:
```diff
@@ -0,0 +1,164 @@
+  return $q.iconSet.credentialCard.defaultImage;
+});
+
+const credentialImageClass = computed(() => {
+  return [`credential-card-image-${props.size}`];
+});
```
```suggestion
const credentialImageClass = computed(
  () => [`credential-card-image-${props.size}`]);
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/credentialCommon.js:
```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright (c) 2022 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
```suggestion
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/credentialCommon.js:
```diff
@@ -0,0 +1,56 @@
+ * Common utilities for credential components.
+ *
+ * @param {object} options - The options to use.
+ * @param {Ref} options.credential - A vue ref to the credential.
```
See lint warning here.

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on components/credentialCommon.js:
```diff
@@ -0,0 +1,56 @@
+
+  const credentialImage = computed(() => {
+    const {image = null, issuer} = unref(credential);
+    if(image) {
+      return image;
+    }
+    if(issuer.image || issuer.logo) {
+      return issuer.image ?? issuer.logo;
+    }
+    return '';
```
Can this be simplified to?:

```suggestion
    return image ?? issuer.image ?? issuer.logo ?? '';
```

tolson4 replied:
4a1122145b9b4b1c9023ab51062930e14c2fe365

dlongley commented on lib/config.js:
```diff
@@ -0,0 +1,13 @@
+import {config} from '@bedrock/web';
+
+config.credentialDisplay = {
+  registration: [/*
```
Hmm, I haven't looked that closely at the config registration format, but if we've just got some array here we may (or may not) need to modify to a map structure to make deployment configuration changes easier. We can leave this for now, just noting that it may come up later and require a modeling change. This is something we can chat with @mattcollier about later if necessary.

dlongley commented on lib/index.js:
```diff
@@ -0,0 +1,25 @@
+export {default as CredentialSwitch} from '../components/CredentialSwitch.vue';
+
+export const registerComponent = async ({app, component, name = ''}) => {
+  app.component(`${name}`, component);
```
Are we expecting `name` to not be a string? If not:

```suggestion
  app.component(name, component);
```

I would think `name` should only ever be a string (or it would be an error).

tolson4 replied:
9801dd68bcde6a1ad29042a986b1752c596b4562

dlongley commented on test/components/TestHome.vue:
```diff
@@ -0,0 +1,52 @@
+import {
+  CredentialSwitch
+} from '@bedrock/vue-vc';
+import {alumniCredential} from './credentials';
+const issuer = 'did:v1:test:1234';
```
Space after imports section:

```suggestion
import {alumniCredential} from './credentials';

const issuer = 'did:v1:test:1234';
```

dlongley reviewed (changes requested):
Thanks! Mostly just style nits in comments here, but a few other important best practice things to get in here.

dlongley commented:
I also couldn't easily tell if this was adding a `.gitignore` and all that stuff. I may have just overlooked it. If so, ignore this, if not, take a look at, e.g., bedrock-vue-wallet to make sure the usual infrastructure files like that are here.

tolson4 committed:
Address several review comments around style.

tolson4 committed:
Update LICENSE.md.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

tolson4 committed:
Use native `Map` in selection logic.

Removes lodash as a dependency.

davidlehn committed:
Remove unused bedrock flag.

davidlehn committed:
Move dependencies.

davidlehn committed:
Enable testing.

tolson4 committed:
Update exports order.

davidlehn commented:
Is there a reason `test/components/credentials.js` parses a long string to JSON?  Can that just be regular JSON in js so it's readable?

tolson4 committed:
Add space after imports in testing file.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

tolson4 committed:
Update testing app and GH actions workflows.

tolson4 commented:
> I also couldn't easily tell if this was adding a `.gitignore` and all that stuff. I may have just overlooked it. If so, ignore this, if not, take a look at, e.g., bedrock-vue-wallet to make sure the usual infrastructure files like that are here.

It's in there - here's a link if you wanted to take a look at it directly: https://github.com/digitalbazaar/bedrock-vue-vc/blob/initial/.gitignore



> Is there a reason `test/components/credentials.js` parses a long string to JSON? Can that just be regular JSON in js so it's readable?

If memory serves right it was done that way so you could have multiple credentials inside that `credentials.js` file. It probably makes more sense to just have them as separate json files grouped in a folder though -> 7d1067c36fc326d38ac69cd676f4599ab88736af

dlongley commented on components/CredentialSwitch.vue:
```diff
@@ -37,49 +39,47 @@ const props = defineProps({
-        map[type] = combined;
+      // if the type already exists in the map, merge it with the
+      // new information
+      if(map.has(type)) {
+        const combined = extend({
+          target: {}, source: [map.get(type), components], deep: true
```
The present implementation hits the map twice, instead hit it just once:

```suggestion
      const existing = map.get(type);
      if(existing) {
        const combined = extend({
          target: {}, source: [existing, components], deep: true
```

dlongley reviewed (approved):
Thanks! One more nit in there ... and we should merge #2 into the initial branch before merging this one.

tolson4 committed:
Update selection logic to not hit map unnecessarily.

Co-authored-by: Dave Longley <dlongley@digitalbazaar.com>

davidlehn committed:
Add npmrc with legacy-peer-deps option.

aljones15 committed:
Add aliases for vue & @bedrock/web.

aljones15 committed:
Fix renderCredential.

tolson4 committed:
Update and fix tests.

tolson4 committed:
Lint.

davidlehn committed:
Lint all source files.

davidlehn committed:
Fix lint issues.

davidlehn committed:
Improve aliases.

- Fix alias matching pattern.
- Remove direct package file accesses.
  - Avoids assuming internal package structure.
  - Let's ESM versions be used automatically.
- Update vue and quasar imports.

davidlehn committed:
Disable coverage.

- vue support not currently working.

davidlehn committed:
Simplify import path.
